zip-0155.rst

```diff
@@ -0,0 +1,246 @@
+ZIP 200. [#zip-0200]_
+
+The terms "Testnet" and "Mainnet" are to be interpreted as described in
+section 3.11 of the Zcash Protocol Specification [#protocol-networks]_.
+
+"P2P network" means the Zcash peer-to-peer network.
+
+
+Abstract
+========
+
+This document proposes a new P2P message to gossip longer node addresses over the
+P2P network. This is required to support new-generation Onion addresses, I2P, and
+potentially other networks that have longer endpoint addresses than fit in the 128
+bits of the current ``addr`` message.
+
+
+Motivation
+==========
+
+Tor v3 hidden services are part of the stable release of Tor since version 0.3.2.9.
+They have various advantages compared to the old hidden services, among which better
+encryption and privacy [#Tor-rendezvous-v3]_. These services have 256-bit addresses
+and thus do not fit in the existing ``addr`` message, which encapsulates onion
+addresses in OnionCat IPv6 addresses.
+
+Other transport-layer protocols such as I2P have always used longer addresses. This
+change would make it possible to gossip such addresses over the P2P network, so that
+other peers can connect to them.
+
+
+Specification
+=============
+
+The ``addrv2`` message is defined as a message where ``pchCommand == "addrv2"``.
+It is  serialized in the standard encoding for P2P messages. Its format is similar
+to the  current ``addr`` message format described in [#Bitcoin-addr-message]_, with
+the difference that the fixed 16-byte IP address is replaced by a network ID and a
+variable-length address, and the services format has been changed to
+[#Bitcoin-CompactSize]_.
+
+This means that the message contains a serialized vector of the following structure:
+
+  +--------------+-----------------+---------------------+----------------------------------------------------------------+
+  | Bytes        | Name            | Data Type           | Description                                                    |
+  +--------------+-----------------+---------------------+----------------------------------------------------------------+
+  | 4            | ``time``        | ``uint32``          | Time that this node was last seen as connected to the network. |
+  |              |                 |                     | A time in Unix epoch time format.                              |
+  +--------------+-----------------+---------------------+----------------------------------------------------------------+
+  | ``varies``   | ``services``    | ``CompactSize``     | Service bits. A CompactSize-encoded bit field that is 64 bits  |
+  |              |                 |                     | wide.                                                          |
+  +--------------+-----------------+---------------------+----------------------------------------------------------------+
+  | 1            | ``networkID``   | ``uint8``           | Network identifier. An 8-bit value that specifies which        |
```

Comment:
![uint8](https://user-images.githubusercontent.com/640955/129425875-efc457da-beaa-41a5-bcf2-f23fd1c4640c.png)

[BIP155](https://github.com/bitcoin/bips/blame/master/bip-0155.mediawiki#L65) specifies `uint8_t` for this field.   I recognize that this is a language specific type. 

I also recognize that `uint8` is not (according to my `clang++ -std=c++17`) a standard type.  

I considered two options:

1.  We could specify that we're implementing against a particular C standard, and use notation derived from that implementation detail.
2.  We could choose type signifiers that are less easily confused with the implementation language syntax.

I am not familiar enough with the ZIP standard to know if it offers guidance.

Reply:
``uint8`` is the convention we use generally in Zcash specifications (e.g. [ZIP 225](https://zips.z.cash/zip-0225#transaction-format) or the protocol spec). I'll add it to the Terminology section.